### ISSUE TITLE
make the EPUB prettier

### ISSUE THREAD

**Issue by anarcat** [wallabag/wallabag#2821]
### Issue details

Now that I have a nice EPUB importer for my ebook reader ([wallabako](https://gitlab.com/anarcat/wallabako/)), I am getting EPUBs from Wallabag, and it's awesome. Thanks for that feature so much!!

They do look a bit ... well, they're ugly. :) The first page says:

> Produced by wallabag with PHPePub
>
> Please open [an issue](https://github.com/wallabag/wallabag/issues) if you have trouble with the display of this E-Book on your device.

This could be much better. First off, this text is great, but should be at the end of the ebook, not the beginning.

Second, the first page should be the title of the book, with the author. Right now, the author is always "wallabag" - how about putting the domain name in there or some other metadata?

Third, a table of contents would be nice. We could parse `<H[1-9]>` tags in the document and generate a table of contents with that?

Finally, how about some images? We do have some image data for articles, we display it on the website - that would make for an awesome cover page!

So, in short:

- [ ] put debug notes at the end
- [ ] set epub author to the article domain or some other metadata
- [ ] generate table of contents from the headings in the document
- [ ] add a cover page based on the article main image
- [ ] `sudo make unicorn` ;)

Thanks!

### Environment

* wallabag version (or git revision) that exhibits the issue: 2.1.5
* How did you install wallabag? Via `git clone` or by downloading the package? unknown
* Last wallabag version that did not exhibit the issue (if applicable): N/A
* php version: unkonwn
* OS: Debian GNU/Linux
* type of hosting (shared or dedicated): dedicated
* which storage system you choose at install (SQLite, MySQL/MariaDB or PostgreSQL): N/A

### Steps to reproduce/test case

Download an EPUB version. Look at it. Just look at it! ;)

**Comment by Quent-in** [comment]
Could we also externalize the text strings so it can be translated ?
To me, it will help epubs to look prettier :)

**Comment by j0k3r** [comment]
@Quent-in of course!

**Comment by pmartin** [comment]
I was about to create an issue, but I think it would be kind of a duplicate of this one and it's just my opinion and not an "issue", so posting here:

 * The 1st page of EPUB exports ("produced by wallabag") is not really *useful*: I have to skip it each time I open an EPUB and never read it as it's always the same. It could be either removed, or placed at the end of the EPUB file, +1 for me.
 * The 2nd page of EPUB exports ("table of contents") is not really *useful* either: it always contains 2 (and only 2) links: "table of contents" and the title of the article. It could be removed too, I think.

I don't think a table of contents is useful (even if improved compared to the existing one), as articles read from wallabag are usually quite short: I pretty much never need to jump to a specific section, while I always have to skip this page before reading the content of the post.
In addition: ereader devices and software often have a "table of contents" feature -- which is probably based on metadata in the EPUB and doesn't require a TOC page present in the book.

I would also not add a page for title + author, as the user would always have to skip this page, which would often not provide much additional information (as title / author are often displayed by the ereader / software, based on metatada in the EPUB).

Basically: when opening an EPUB, my goal is to read the *content* of the article -- and it would be more user-friendly if I didn't have to navigate at all before reading.
Would that fit into what you were thinking about?

**Comment by anarcat** [comment]
i agree with you. the first page is the "put debug notes at the end" task. the table of contents issue is "generate table of contents from the headings in the document" - which i think would make it actually useful. some ebooks actually *have* those verbatim in the book, so i think it's fine to have it there.

as for the author and other metadata, that would be strictly in the EPUB metadata, not in the text itself, naturally.

**Comment by anarcat** [comment]
while i'm here fantasizing - could we get the estimated reading time as well please? thanks! ;)

**Comment by pieplu** [comment]
Anyone work on "add a cover page based on the article main image and time to read" ?

**Comment by anarcat** [comment]
i cleaned up the todo list to remove the unicorn. i also split off the time to read in a separate entry: there's actual epub metadata for time to read information that can be read by e-readers, on top of including it on the cover page.

@pieplu as far as I know, no one is working on this issue, but hopefully if someone is, they will mention it here.

**Comment by anarcat** [comment]
So an update on this since 2.3 was released: there has been *some* improvements: #3023, #3392 and other changes cleaned up the main page so it shows the title and the TOC is at the end. #3266 fixed the ePUB metadata so it's actually useful as well, which is really nice.

what's still missing is the cover page - i filed #3602 to followup on that. the TOC is also pretty useless, so I opened #3603 to followup on that specifically. the time to read metadata, include URL and image inclusion reliability issues I have not filed as separate issues just yet, as I do not see a clear way forward on those...

thanks for the 2.3 release, it's great! :)

**Comment by Napator** [comment]
Hi, is there a simple way, when exporting several articles to an ePub, to get the title at the top of each one ? 
For now the title of the first article is putted on the first page, but the titles of the others articles are only in the table of contents, not at the top of each article.

It'd be perfect to add the title on top of each article, and why not the author (e.g. website), like when i click on the article in the wallabag web client.

I'd like to test that, but don't have my own instance.

Thank you

**Comment by nicosomb** [comment]
Hello there, 

Today, wallabag sets the ePub date with the creation date of the article (the date when we save the article in wallabag, see https://github.com/wallabag/wallabag/blob/master/src/Wallabag/CoreBundle/Helper/EntriesExport.php#L204)
I think that it should be better if the used date would be the published date of the article. 

Any opinion before I open my PR?

**Comment by Kdecherf** [comment]
@nicosomb I would go with both dates, one line for "added on" and one for "published on"

**Comment by PackElend** [comment]
> * [ ]  include images reliably (not filed)
> *

if Im not mistaken, that is the only remaining open point, isn't it?

**Comment by harmtemolder** [comment]
Looking at the only remaining open point:
 
1. https://ottawacitizen.com/news/national/built-to-crash-the-ugly-sputtering-beginning-of-shared-services-and-how-politics-conspired-against-it fails with the same error as reported in #7477 (= <https://github.com/wallabag/PHPePub/issues/5>)
2. https://www.psychologytoday.com/intl/articles/201611/are-we-forfeiting-control-over-our-digital-lives exports fine, but doesn't show images in calibre's E-book viewer. They do show OK in KOReader on my Android phone, so that's probably a calibre issue.

So this issue could be closed, maybe?
